contrib/win32/win32compat/misc.c

```diff
@@ -1725,3 +1725,177 @@ get_user_sid(char* name)
+
+#define CMDLINE_APPEND(P, S)		\
+do {					\
+	int _S_len = (int)strlen(S);		\
+	memcpy((P), (S), _S_len);	\
+	(P) += _S_len;			\
+} while(0)
+
+	/* get shell type */
+	if (strstr(shell, "system32\\cmd"))
+		shell_type = SH_CMD;
+	else if (strstr(shell, "powershell"))
+		shell_type = SH_PS;
+	else if (strstr(shell, "system32\\bash"))
+		shell_type = SH_WSL_BASH;
+	else if (strstr(shell, "cygwin"))
+		shell_type = SH_CYGWIN;
+
+	/* special case where incoming command needs to be adjusted */
+	do {
+		/*
+		* identify scp and sftp sessions
+		* we want to launch scp and sftp executables from the same binary directory
+		* that sshd is hosted in. This will facilitate hosting and evaluating
+		* multiple versions of OpenSSH at the same time.
+		*
+		* currently we can only accomodate this for cmd.exe, since cmd.exe simply executes 
+		* its commandline without applying CRT or shell specific rules
+		*
+		* Ex.
+		* this works
+		*	cmd /c "c:\program files\sftp" -d
+		* this following wouldn't work in powershell, cygwin's or WSL bash unless we put
+		* some shell specific rules 
+		*	powershell -c "c:\program files\scp" -t
+		*	cygwin\bash -c "c:\program files\scp" -t
+		*	bash -c "c:\program files\scp" -t
+		* 
+		* for now, for all non-cmd shells, we launch scp and sftp-server directly -
+		*	shell -c "scp.exe -t"
+		* note that .exe extension and case matters for WSL bash
+		* note that double quotes matter for WSL and Cygwin bash, they dont matter for PS
+		*
+		* consequence - 
+		* for non-cmd shells - sftp and scp installation path is expected to be in machine wide PATH
+		* 
+		*/
+
+		int command_len;
+		const char *command_args = NULL;
+
+		if (!command)
+			break;
+		command_len = (int)strlen(command);
+
+		if (command_len >= 13 && _memicmp(command, "internal-sftp", 13) == 0) {
+			command_type = CMD_SFTP;
+			command_args = command + 13;
+		} else if (command_len >= 11 && _memicmp(command, "sftp-server", 11) == 0) {
+			command_type = CMD_SFTP;
+
+			/* account for possible .exe extension */
+			if (command_len >= 15 && _memicmp(command + 11, ".exe", 4) == 0)
+				command_args = command + 15;
+			else
+				command_args = command + 11;
+		} else if (command_len >= 3 && _memicmp(command, "scp", 3) == 0) {
+			command_type = CMD_SCP;
+
+			/* account for possible .exe extension */
+			if (command_len >= 7 && _memicmp(command + 3, ".exe", 4) == 0)
+				command_args = command + 7;
+			else
+				command_args = command + 3;
+		}
+
+		if (command_type == CMD_OTHER)
+			break;
+
+		len = 0;
+		len += progdir_len + 4; /* account for " around */
+		len += command_len + 4; /* account for possible .exe addition */
+
+		if ((cmd_sp = malloc(len)) == NULL) {
+			errno = ENOMEM;
+			goto done;
+		}
+
+		p = cmd_sp;
+
+		if (shell_type == SH_CMD) {
+
+			CMDLINE_APPEND(p, "\"");
+			CMDLINE_APPEND(p, progdir);
+
+			if (command_type == CMD_SCP)
+				CMDLINE_APPEND(p, "\\scp.exe\"");
+			else
+				CMDLINE_APPEND(p, "\\sftp-server.exe\"");
+
+		} else {
```

Comment:
For shell_type != SH_CMD, how will other shells know where to find the executable?  Is it expected to be in the system path?  If its depending on the current working directory, might be good to make sure this works with your new chroot() jail since I think that changes the working directory somewhere along the way.

Reply:
Yes, as commented in earlier part of the code (and as will be documented once this change is in), its expected that scp and sftp-server (openssh installation folder) is in system path. Current working directory does not matter.  

@bagajjal  and I discussed an alternative to work around this limitation, i.e to runtime add sshd program directory to path before launching scp and sftp sessions. However, this does not completely guarantee that the sshd associated versions of scp and sftp-server are always picked, based on the info at https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx. 
Also, this behavior is inline with how things work on Unix, where the sshd installation path is expected to be in system path. 

We don't apply chroot while executing child processes.